core/src/oop_module.c

```diff
@@ -0,0 +1,308 @@
+
+#include "oop_module.h"
+#include "broker.h"
+
+#include "stdlib.h"
+
+#include "nanomsg/nn.h"
+#include "nanomsg/reqrep.h"
+
+#include "azure_c_shared_utility/lock.h"
+#include "azure_c_shared_utility/threadapi.h"
+
+#include "message.h"
+#include "module.h"
+
+typedef MESSAGE_THREAD * MESSAGE_THREAD_HANDLE;
+
+int
+connect_to_gateway_with_connection_key (
+    OOP_MODULE_HANDLE oop_module_,
+    const char * connection_key_
+);
+
+int
+invoke_add_module_process (
+    OOP_MODULE_HANDLE oop_module_,
+    const char * json_config_
+);
+
+int
+message_pump_thread (
+    void * thread_arg_
+);
+
+int
+process_create_message_payload (
+    OOP_MODULE_HANDLE oop_module_,
+    const char * create_message_payload_
+);
+
+typedef struct MESSAGE_THREAD_TAG {
+    bool halt;
+    LOCK_HANDLE mutex;
+    THREAD_HANDLE thread;
+} MESSAGE_THREAD;
+
+typedef struct OOP_MODULE_TAG {
+    int command_socket;
+    int message_socket;
+    MESSAGE_THREAD_HANDLE message_thread;
+    MODULE module;
+} OOP_MODULE;
+
+OOP_MODULE_HANDLE
+OopModule_Attach (
+    const MODULE_API * module_api_,
+    const char * const connection_key_
+) {
```

Comment:
I assume error handling for public functions will come in another commit added to this PR...